diagnostics/diagnostic.go

```diff
@@ -468,6 +474,59 @@ func (d *DiagnosticClient) runSnapshotFilesListListener() {
 	}()
 }
 
+func appendWithCap(list *list.List, cap int, items []time.Duration) {
+	if list.Len() == cap {
+		list.Remove(list.Front())
+	}
+
+	for _, v := range items {
+		list.PushBack(v)
+	}
+}
+
+func (d *DiagnosticClient) runBlockMetricsListener() {
+	go func() {
+		ctx, ch, cancel := diaglib.Context[diaglib.AppendBlockMetrics](context.Background(), 1)
+		defer cancel()
+
+		rootCtx, _ := common.RootContext()
+
+		diaglib.StartProviders(ctx, diaglib.TypeOf(diaglib.AppendBlockMetrics{}), log.Root())
+		for {
+			select {
+			case <-rootCtx.Done():
+				cancel()
+				return
+			case info := <-ch:
+				d.mu.Lock()
```

Comment:
would suggest to use `defer` + `func` when working with locks, for example
```
func() {
    d.mu.Lock()
    defer d.mu.Unlock()

    // your code
}()
```

benefit of using defer is that it is guaranteed to be executed once the func exits - even in the case of panics